### PR TITLE
057 bad zero power solution from mixed convection

### LINKED ISSUE #57: Bad results from mixed_convection solver when imposing zero power
The new implementation of the `mixed_convection` formulation introduced in #51 solved the Singular Matrix problem when zero power is imposed.
However, the results of the current implementation are not satisfactory in terms of energy conservation.

By simulating 40 cm of rods having imposed zero power, I found that the temperatures are no longer constant throughout the assembly.
In particular, there is a region of the axial section with lower temperatures that suggests errors in energy conservation.

Image
 
A possible cause of the error could be the new explicit treatment of transverse energy mixing, which was moved from the main matrix definition and included in the vector of known terms (#51 modifications in `_build_vector`).
In the new implementation this term of the energy equation (which is a function of $h^*$, $\Delta \rho$ and $\Delta v$) is calculated only once at the call of the `_build_vector()` method. This is done before entering the loop in which $\Delta \rho$ and $\Delta v$ are calculated.

This procedure introduces inaccuracies into the energy balance of the i-th axial layer, because the new additional source term is calculated using only information from the (i-1)-th layer.

To correctly satisfy the energy balance, the vector of known terms must be updated at each iteration step using the newly computed values of $\Delta \rho$ and $\Delta v$.

### PR DESCRIPTION
**Pull request by Nino-ndrk**
Improvements to energy conservation in mixed convection solver #57

### CODE REVIEW

File: dassh/region_mixed.py [hunk 1]
```diff
@@ -578,6 +578,7 @@ def _build_matrix(self, dz: float, delta_v: np.ndarray,
             Coefficient matrix for the system to solve
         """
         self._vstar = self._calc_star_quantity(delta_v, delta_rho, nn, 'v')
```
Comment (lelaus) [inline dassh/region_mixed.py]: I understand the need for calculating `self._hstar` at each iteration before updating the variable part of the known vector. In this way, the known vector is always aligned with the results of the previous iteration. However, to improve readability, I suggest moving this calculation before calling the method `_build_matrix()` as the term `self._hstar` is not used here, but in the iterative solving loop. The same goes for `self._vstar` even though it is used directly in `_build_matrix()`: for consistency they could both be moved before calling `_build_matrix()`. This appears not to imply any other modification in the code.

File: tests/test_region_mixed.py
Comment (lelaus) [inline tests/test_region_mixed.py]: According to the needs covered by this test, and to the modifications applied to the code, I suggest replacing "the Delta(mh) term is equal to zero" with "the mass balance is ensured and no enthalpy variation is calculated."

File: tests/test_region_mixed.py [hunk 2]
```diff
@@ -348,4 +352,34 @@ def test_calc_star_quantities(self,
...
             rr_data.mixed['deltarho'], rr_data.mixed['RR_star_test'], 
             expected_hstar, expected_vstar)
-        
\ No newline at end of file
+
+
+def test_zero_power_enthalpy_jump(testdir: str):
+    """
+    Test the mixed_convection solver in isothermal case (rods with zero
+    power deposition). Check that relative error in global enthalpy jump is
```
Comment (lelaus) [inline tests/test_region_mixed.py]: For precision, I suggest replacing "rods with zero power deposition" with "no power entering the coolant, either as a source term or as a heat flux from the external duct" as the power assigned to the rods is not the only way to provide heat to the coolant.

**Review by lelaus — CHANGES_REQUESTED** [review]

File: dassh/region_mixed.py (hunk 1 above)
Comment (Nino-ndrk) [reply dassh/region_mixed.py]: Solved in 005a6e2

File: tests/test_region_mixed.py
Comment (Nino-ndrk) [reply tests/test_region_mixed.py]: Solved in 005a6e2

File: tests/test_region_mixed.py (hunk 2 above)
Comment (Nino-ndrk) [reply tests/test_region_mixed.py]: Solved in 005a6e2

**Review by lelaus — APPROVED** [review]
All my comments have been managed satisfactorily.
